Src/GoogleApis.Core/Apis/Util/PageStreamer.cs

```diff
@@ -0,0 +1,137 @@
+        /// into a new request. Must not be null.</param>
+        /// <param name="tokenExtractor">Function to extract the next page token from a response.
+        /// Must not be null.</param>
+        /// <param name="resourceExtractor">Function to extract a sequence of resources from a response.
+        /// Must not be null.</param>
+        /// <param name="emptyToken">The token to use to detect when there are no more pages available.</param>
+        public PageStreamer(
+            Func<TRequest, TToken, TRequest> requestProvider,
+            Func<TResponse, TToken> tokenExtractor,
+            Func<TResponse, IEnumerable<TResource>> resourceExtractor,
+            TToken emptyToken)
+        {
+            if (requestProvider == null)
+            {
+                throw new ArgumentNullException("requestProvider");
+            }
+            if (tokenExtractor == null)
+            {
+                throw new ArgumentNullException("tokenExtractor");
+            }
+            if (resourceExtractor == null)
+            {
+                throw new ArgumentNullException("resourceExtractor");
+            }
+            this._requestProvider = requestProvider;
+            this._tokenExtractor = tokenExtractor;
+            this._resourceExtractor = resourceExtractor;
+            this._emptyToken = emptyToken;
+        }
+
+        /// <summary>
+        /// Lazily fetches resources a page at a time.
+        /// </summary>
+        /// <param name="initialRequest">The initial request to send. If this contains a page token,
+        /// that token is maintained. This is also passed to the request provider along with a token
+        /// to provide another request. (The provider may build a new object, or modify the existing request.)</param>
+        /// <param name="responseFetcher">A function to fetch a response containing a page of results given a request.</param>
+        /// <returns>A sequence of resources, which are fetched a page at a time. Must not be null.</returns>
+        public IEnumerable<TResource> Fetch(TRequest initialRequest, Func<TRequest, TResponse> responseFetcher)
+        {
+            if (responseFetcher == null)
+            {
+                throw new ArgumentNullException("responseFetcher");
+            }
+            TToken token;
+            TRequest nextRequest = initialRequest;
+            do
+            {
+                TResponse response = responseFetcher(nextRequest);
+                token = _tokenExtractor(response);
+                nextRequest = _requestProvider(initialRequest, token);
+                foreach (var item in _resourceExtractor(response))
+                {
+                    yield return item;
+                }
+            } while (!IsEmptyToken(token));
+        }
+
+        /// <summary>
+        /// Asynchronously (but asynchronously) fetches a complete set of resources, potentially making multiple requests.
```

Comment:
Only my brain :) Fixed in another commit. (Will squash all of these, of course.)